extmod/machine_usb_device.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2024 Angus Gratton
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/mpconfig.h"
+
+#if MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE
+
+#include "mp_usbd.h"
+#include "py/mperrno.h"
+#include "py/objstr.h"
+
+// Implements the singleton runtime USB object
+//
+// Currently this implementation references TinyUSB directly.
+
+#ifndef NO_QSTR
+#include "device/usbd_pvt.h"
+#endif
+
+#define HAS_BUILTIN_DRIVERS (MICROPY_HW_USB_CDC || MICROPY_HW_USB_MSC)
+
+const mp_obj_type_t machine_usb_device_type;
+
+static mp_obj_t usb_device_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    (void)type;
+    (void)n_args;
+    (void)n_kw;
+    (void)args;
+
+    if (MP_STATE_VM(usbd) == MP_OBJ_NULL) {
+        mp_obj_usb_device_t *o = m_new0(mp_obj_usb_device_t, 1);
+        o->base.type = &machine_usb_device_type;
+        o->desc_dev = mp_const_none;
+        o->desc_cfg = mp_const_none;
+        o->desc_strs = mp_const_none;
+        o->open_itf_cb = mp_const_none;
+        o->reset_cb = mp_const_none;
+        o->control_xfer_cb = mp_const_none;
+        o->xfer_cb = mp_const_none;
+        for (int i = 0; i < CFG_TUD_ENDPPOINT_MAX; i++) {
+            o->xfer_data[i][0] = mp_const_none;
+            o->xfer_data[i][1] = mp_const_none;
+        }
+        o->builtin_driver = MP_OBJ_TO_PTR(&mp_type_usb_device_builtin_none);
```

Comment:
This will need to be `MP_OBJ_FROM_PTR`, because you want to create an object.

Reply:
Oops, I think that'll be the case in a few places...